source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,315 @@
+     * Collective kDNSServiceFlagsMoreComing flag:
+     * When [DNSServiceGetAddrInfoReply] are invoked using a shared DNSServiceRef, the
+     * kDNSServiceFlagsMoreComing flag applies collectively to *all* active
+     * operations sharing the same [main_sd_ref]. If the MoreComing flag is
+     * set it means that there are more results queued on this parent DNSServiceRef,
+     * but not necessarily more results for this particular callback function.
+     * The implication of this for client programmers is that when a callback
+     * is invoked with the MoreComing flag set, the code should update its
+     * internal data structures with the new result (as is done above when calling
+     * parent_.addPendingQuery(this))...Then, later when a callback is eventually invoked with the
+     * MoreComing flag not set, the code should update *all* [pending queries] related to that
+     * shared parent DNSServiceRef that need updating (i.e that have had DNSServiceGetAddrInfoReply
+     * called on them since the last flush), not just the [queries] related to the particular
+     * callback that happened to be the last one to be invoked.
+     */
+    ENVOY_LOG(warn, "DNS Resolver flushing queries pending callback");
+    parent_.flushPendingQueries();
+  }
+}
+
+DNSServiceErrorType
+AppleDnsResolverImpl::PendingResolution::dnsServiceGetAddrInfo(DnsLookupFamily dns_lookup_family) {
+  DNSServiceProtocol protocol;
+  switch (dns_lookup_family) {
+  case DnsLookupFamily::V4Only:
+    protocol = kDNSServiceProtocol_IPv4;
+    break;
+  case DnsLookupFamily::V6Only:
+    protocol = kDNSServiceProtocol_IPv6;
+    break;
+  case DnsLookupFamily::Auto:
+    protocol = kDNSServiceProtocol_IPv4 | kDNSServiceProtocol_IPv6;
+    break;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+
+  // TODO: explore caching: there are caching flags in the dns_sd.h flags, allow expired answers
+  // from the cache?
+  // TODO: explore validation via DNSSEC?
+  return DNSServiceGetAddrInfo(
+      &individual_sd_ref_, kDNSServiceFlagsShareConnection | kDNSServiceFlagsTimeout, 0, protocol,
+      dns_name_.c_str(),
+      /*
+       * About Thread Safety (taken from inline documentation there):
+       * The dns_sd.h API does not presuppose any particular threading model, and consequently
+       * does no locking internally (which would require linking with a specific threading library).
+       * If the client concurrently, from multiple threads (or contexts), calls API routines using
+       * the same DNSServiceRef, it is the client's responsibility to provide mutual exclusion for
+       * that DNSServiceRef.
+       */
+
+      // Therefore, much like the c-ares implementation All calls and callbacks to the API need to
+      // happen on the thread that owns the creating dispatcher. This is the case as callbacks are
+      // driven by processing bytes in onEventCallback which run on the passed in dispatcher's event
+      // loop.
```

Comment:
If this is the case I don't think you need any posts?

Reply:
yep, and also exceptions seem to be handled correctly within the context of the library given the tests.